numpy/lib/npyio.py

```diff
@@ -1459,13 +1459,13 @@ def fromregex(file, regexp, dtype, encoding=None):
             dtype = np.dtype(dtype)
 
         content = file.read()
-        if isinstance(content, bytes) and not isinstance(regexp, bytes):
-            regexp = asbytes(regexp)
-        elif not isinstance(content, bytes) and isinstance(regexp, bytes):
-            regexp = asstr(regexp)
```

Comment:
Seems like maybe this should just replace `not isinstance(content, bytes)` with `isinstance(content, np.unicode)`, and then could remain where it is.

Reply:
`not isinstance(content, bytes)` is not at fault here. The problem is checking  `not isinstance(regexp, bytes)` on regular expression objects (i.e. already compiled).

Reply:
Sorry, that was what I meant to write. I'd argue the problem is that we're assuming that everything which is not bytes must be unicode, which isn't reasonable.